src/auth/impersonated.ts

```diff
@@ -0,0 +1,194 @@
+  targetPrincipal?: string;
+  targetScopes?: string[];
+  delegates?: string[];
+  lifetime?: number | 3600;
+  endpoint?: string | 'https://iamcredentials.googleapis.com';
+}
+
+export interface TokenResponse {
+  accessToken: string;
+  expireTime: string;
+}
+
+export class Impersonated extends OAuth2Client {
+  private sourceClient: OAuth2Client;
+  private targetPrincipal: string;
+  private targetScopes: string[];
+  private delegates: string[];
+  private lifetime: number;
+  private endpoint: string;
+
+  /**
+   * Impersonated service account credentials.
+   *
+   * Create a new access token by impersonating another service account.
+   *
+   * Impersonated Credentials allowing credentials issued to a user or
+   * service account to impersonate another. The source project using
+   * Impersonated Credentials must enable the "IAMCredentials" API.
+   * Also, the target service account must grant the orginating principal
+   * the "Service Account Token Creator" IAM role.
+   *
+   * @param credentials the service account email address.
+   * @param sourceClient the source credential used as to acquire the
+   * impersonated credentials
+   * @param targetPrincipal the service account to impersonate.
+   * @param delegates the chained list of delegates required to grant the
+   *  final access_token. If set, the sequence of identities must have
+   * "Service Account Token Creator" capability granted to the preceding
+   * identity. For example, if set to [serviceAccountB, serviceAccountC],
+   * the sourceCredential must have the Token Creator role on serviceAccountB.
+   * serviceAccountB must have the Token Creator on serviceAccountC.
+   * Finally, C must have Token Creator on target_principal.
+   * If left unset, sourceCredential must have that role on targetPrincipal.
+   * @param targetScopes scopes to request during the authorization grant.
+   * @param lifetime number of seconds the delegated credential should be
+   * valid for (up to 3600).
+   * @param endpoint api endpoint override.
+   */
+  constructor(options: ImpersonatedOptions = {}) {
+    super(options);
+    this.credentials = {
+      expiry_date: 1,
+      refresh_token: 'impersonated-placeholder',
+    };
+    this.sourceClient = options.sourceClient || new OAuth2Client();
+    this.targetPrincipal = options.targetPrincipal || '';
+    this.delegates = options.delegates || [];
+    this.targetScopes = options.targetScopes || [];
+    this.lifetime = options.lifetime || 3600;
+    this.endpoint = options.endpoint || 'https://iamcredentials.googleapis.com';
+  }
+
+  /**
+   * Refreshes the access token.
+   * @param refreshToken Unused parameter
+   */
+  protected async refreshToken(
+    refreshToken?: string | null
+  ): Promise<GetTokenResponse> {
+    const iat = Date.now();
+    if (this.credentials.expiry_date) {
+      if (this.credentials.expiry_date <= iat) {
+        try {
+          await this.sourceClient.getAccessToken();
```

Comment:
I switched from `.then`, `.catch` to `async`/`await` which is more idiomatic with how we tend to write JavaScript.